overlord/state/change_test.go

```diff
@@ -113,3 +113,67 @@ func (cs *changeSuite) TestTasksNeedsLocked(c *C) {
 
 	c.Assert(func() { chg.Tasks() }, PanicMatches, "internal error: accessing state without lock")
 }
+
+func (cs *changeSuite) TestStatusAndSetStatus(c *C) {
+	st := state.New(nil)
+	st.Lock()
+	defer st.Unlock()
+
+	chg := st.NewChange("install", "...")
+
+	// default with no tasks will end up as Done
+	c.Check(chg.Status(), Equals, state.Done)
+
+	chg.SetStatus(state.Running)
+
+	c.Check(chg.Status(), Equals, state.Running)
+}
+
+func (cs *changeSuite) TestStatusNeedsLock(c *C) {
+	st := state.New(nil)
+	st.Lock()
+	chg := st.NewChange("install", "...")
+	st.Unlock()
+
+	c.Assert(func() { chg.Status() }, PanicMatches, "internal error: accessing state without lock")
+}
+
+func (cs *changeSuite) TestSetStatusNeedsLock(c *C) {
+	st := state.New(nil)
+	st.Lock()
+	chg := st.NewChange("install", "...")
+	st.Unlock()
+
+	c.Assert(func() { chg.SetStatus(state.Waiting) }, PanicMatches, "internal error: accessing state without lock")
+}
+
+func (cs *changeSuite) TestStatusDerivedFromTasks(c *C) {
+	st := state.New(nil)
+	st.Lock()
+	defer st.Unlock()
+
+	chg := st.NewChange("install", "...")
+
+	t1 := chg.NewTask("download", "1...")
+	t2 := chg.NewTask("verify", "2...")
+
+	c.Check(chg.Status(), Equals, state.Running)
+
+	t1.SetStatus(state.Waiting)
+	c.Check(chg.Status(), Equals, state.Running)
+
+	t2.SetStatus(state.Waiting)
+	c.Check(chg.Status(), Equals, state.Waiting)
+
+	t1.SetStatus(state.Error)
```

Comment:
@zyga_pl I think the thinking that after a while all tasks will either converge to Error or Done
 at that point the full Change will be Error

[3:21:34 PM] Zygmunt Krynicki:
mmm, that's true
 +1 then

[3:22:13 PM] Samuele Pedroni:
anyway doing another PR and then getting back to the small fixes in this one
 an orthogonal PR to be clear

[3:23:34 PM] Gustavo Niemeyer:
RIght, and if something is still waiting or running, that's more important than knowing that something else has failed

[3:24:10 PM] Samuele Pedroni:
yes, means the change isn't in a steady-state yet
 the change as a whole